### PR TITLE
allow admin to enforce password on mail shares

### PR DESCRIPTION
**Pull request by schiessle**
Allow the admin to enforce passwords on share by mails.

That's my conclusion out if this discussion for now: https://github.com/nextcloud/server/issues/2357#issuecomment-292938422

### Scenario 1:

Admin checks the "enforce password" in the mail share settings, everything else stays the default:

- The user creates a share
- We create a random password on the server and set it for the share
- We send out the share
- We send out a separate mail with the random password to the recipient

Users can change the password at any time but they can't disable it. Changing the password will trigger a new mail to the recipient with the new password.

### Scenario 2

Admin checks "enforce password" in the mail share settings and disables the password mail:

- The user creates a share
- We create a random password on the server and set it for the share
- We send out the share
- We send a mail to _the user who just created the share_. Tell him that according to the security policy on his server all shares needs to be password protected and tell him the auto-generated password. Now it is the job of the user to get the password to the recipient.

Users can change the password at any time but they can't disable it, no additional mail will be send.

### Scenario 3

Admin checks "enforce password" in the mail share settings, disables the password mail and the user who creates the share has no email address (really unlikely combination but it can still happen):

- The user creates a share
- operation fails, and the user gets a warning that according to the security policy on the server he needs a valid email address to create a mail share. He should set one in the personal settings and try again.

I like the workflow, especially in scenario 1 which is probably ~70% or all use cases, maybe 29% will have a setup as in scenario 2 which I think works still quite well and probably around 1% will be confronted with scenario 3. In this rare case you could even argue that it is good to remember people to set a valid email address because they will also need it in case they lose their password or for other tasks.

Reviews? Opinions?

Note: that I don't have a lot of extra time I can put into this feature. Complete re-designs are out-of-scope for now. Either we take this one with small adjustments or nothing for Nc12 (at least nothing I implement 😉 )

cc @rullzer @MorrisJobke @jancborchardt

### CODE REVIEW

**Comment by ChristophWurst** [comment]
conflicting

**Comment by schiessle** [comment]
I just did a rebase and resolved the merge conflicts but couldn't test it afterwards. Hope that nothing breaks... Will test tomorrow.

File: apps/sharebymail/appinfo/app.php [hunk 1]
```diff
@@ -19,6 +19,7 @@
  *
  */
 
-$settings = new \OCA\ShareByMail\Settings();
+$settings = new \OCA\ShareByMail\Settings(new \OCA\ShareByMail\Settings\SettingsManager(\OC::$server->getConfig()));
```
Comment (schiessle) [inline apps/sharebymail/appinfo/app.php]: done

**Comment by MorrisJobke** [comment]
@schiessle Could you please fix the emails, thanks :)

**Comment by schiessle** [comment]
changed to new mail template... please review. Thanks!

**Comment by supremesyntax** [comment]
I have checked out the branch and tested it a minute ago. Looks good to me great work.
Though i can't see anything in admin>Security section... Does anyone else have this problem for this branch?

**Comment by MorrisJobke** [comment]
> Though i can't see anything in admin>Security section - just a blank page. Does anyone else have this problem for this branch?

You need the bruteforcesettings app cloned in you apps folder. It is shipped in the release tar ball and automatically enabled. ;)

**Comment by supremesyntax** [comment]
@MorrisJobke Thanks for the hint :D I can't find the password policy settings, thought they might be shifted from additional settings to security but i was wrong... do i need some other apps enabled? Have these settings been removed for NC12?
Right now the auto-generated password is 8 chars upper-/lower-case but i have advanced requirements.

**Comment by schiessle** [comment]
> You need the bruteforcesettings app cloned in you apps folder. It is shipped in the release tar ball and automatically enabled. ;)

But of course we shouldn't show a empty security section, even if the app is shipped and enabled by default. Admins can disable it at any time. @supremesyntax can you create a separate issue for it? Thank.

**Comment by schiessle** [comment]
@rullzer @MorrisJobke @nickvergessen @ChristophWurst can we please do a final review and get this in? I don't want to rebase and adjust the code over and over again (which will happen again one https://github.com/nextcloud/server/pull/4384 is merged)

Thanks a lot! 😃

**Comment by MorrisJobke** [comment]
> @rullzer @MorrisJobke @nickvergessen @ChristophWurst can we please do a final review and get this in? I don't want to rebase, adjust the code and test it over and over again (which will happen again once #4384 is merged)

Don't worry - I fixed it for you 😉 Let me test it now.

**Review by MorrisJobke — APPROVED** [review]
Rebased, fixed a typo, and tested 👍 Works

**Comment by supremesyntax** [comment]
Could somebody please answer my question? Maybe y'all overlooked it :smile: 
I want to test if the auto-generated password respects the password policy...

> I can't find the password policy settings, thought they might have been shifted from additional settings to security but i was wrong... do i need some other apps enabled? Have these settings been removed for NC12?

**Comment by rullzer** [comment]
@supremesyntax if you have the password policy app enabled they should show up (in 12 they will be moved to the security settings btw).

**Review by rullzer — APPROVED** [review]
Ok lets do this!

**Comment by schiessle** [comment]
@supremesyntax  it will, but only with the password policy app we will release together with Nextcloud 12, the current version in Nextcloud 11 doesn't expose the policy to other apps.

**Pull request merged**